File: C/cp/maxSubArraySum.c
#include <stdio.h>
#include <stdlib.h>
#define MAX(a,b) (((a)>(b))?(a):(b))

int main(void) {
    
	// Accepting an array 
	int size;
	printf("Enter the size of the array ");
	scanf ("%d", &size);
	
	printf("Enter elements of array \n");
	int arr[size];
	for (int index = 0; index < size; index++ ){
	    scanf("%d", &arr[index]);
	}
	
	//KADANE'S ALGORITHM
	int max_sum  = arr[0];
	int curr_max = arr[0];
	for (int index = 1; index < size; index++ ){
	    curr_max = MAX(arr[index], arr[index] + curr_max);
	    max_sum = MAX(curr_max, max_sum);
	}
	
	printf("Maximum sum of subarray is %d \n", max_sum);
	
	return 0;
}

/*
TIME COMPLEXITY : O(n) where n is number of elements in array
SPACE COMPLEXITY : O(1) as no new space is allocated
TEST CASE 

INPUT
Enter number of elemets in array 4
Enter elemets of array  2 -3 4 6 

OUTPUT
Maximum sum of subarray is 10 


INPUT
Enter number of elemets in array 4
Enter elemets of array  2 3 4 6 

OUTPUT
Maximum sum of subarray is 15 
*/

File: Python/other/MaxSubArraySum.py
# function to find maximum subarray sum
# using Kadane's algorithm
def max_subarray_sum(arr, n):
    final_max = arr[0]
    curr_max = arr[0]

    for i in range(1, n):
        curr_max = max(arr[i], curr_max + arr[i])
        final_max = max(final_max, curr_max)

    return final_max

def main():
    arr = [-13, -3, -25, -20, -3, -16, -23, -12, -5, -22, -15, -4, -7]
    max_sum = max_subarray_sum(arr, len(arr))
    print(f"The given array: {arr}")
    print(f"The maximum subarray sum is {max_sum}")

if __name__ == '__main__':
    main()